Automattic/jetpack
# Premium Content blocks: Fix parent block selector button in FSE and when block is nested

Related to: #19504

I noticed while testing the Premium Content blocks in the full site editor that the parent block selector button doesn't work properly in FSE. If you insert a Premium Content block into the Footer template part, then select a heading or paragraph within the Premium Content block, and then use the parent block selector button, the Footer template part will be selected, instead of the Premium Content container.

It turns out that the block had been using the highest root parent for its behaviour where selecting the logged out or subscriber view blocks moves the selection to the premium content container (since it manages selection between the two views). Usually, in a non-FSE site, the premium content block will be inserted at the root level, so we didn't notice the bug up until now. However, in FSE, the block is usually a child of a template part block, so the bug becomes apparent.

The fix is to use similar logic to the parent block selector button itself, and grab the immediate parent client id.

#### Changes proposed in this Pull Request:

* In the premium content subscriber and logged-out views, update the parent client id to use the nearest parent to the block, instead of the root block
* Add a couple of short comments to make it clear why we're managing selection in these blocks

##### Before

Note that selecting the parent moves the selection to the footer template part instead of to the premium content container.

![parent-selector-before-sml](https://user-images.githubusercontent.com/14988353/116658152-e3806180-a9d2-11eb-94a7-aa0368ff8277.gif)

##### After

Selection is correctly moved to the parent container.

![parent-selector-after-sml](https://user-images.githubusercontent.com/14988353/116658249-06127a80-a9d3-11eb-9c79-81dbaba0d483.gif)

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






##### Test in FSE

* Create a site with an FSE theme, e.g. via https://horizon.wordpress.com/new?flags=gutenboarding/site-editor
* Before applying this change insert a premium content block into the footer template part
* Click on a heading or paragraph within the block to select it
* Then, click the parent block selector button (the top left button in the toolbar)
* Notice that the selection is moved to the footer template part, instead of to the premium content container
* Apply this change
* Reload the editor, and repeat, noting that the selection should be moved from the heading or paragraph block up to the premium content container block
* Try out the Subscriber and Visitor view toggles and check that the fix works for both kinds of views

##### Test in non-FSE

* In a non-FSE site, within the post editor, create some nested blocks (e.g. a Group block with a Columns block within it, and Column blocks within that).
* Within the nested blocks, insert a Premium Content block
* Before applying this change, note that the same bug is present, where using the parent block selector will select the root block (e.g. the Group block) instead of the premium content container
* Apply the change and reload the editor
* The parent block selector button should now move the selection from the paragraph or heading blocks to the premium content container
